crates/cacache/src/put.rs

```diff
@@ -47,6 +47,165 @@ where
     writer.commit().await
 }
 
+
+pub trait FileLike : Read {
+    fn path(&self) -> Result<String>;
+    fn size(&self) -> usize;
+    fn mode(&self) -> Result<u32>;
+}
+
+struct SSRIStream<Inner> {
+    builder: ssri::IntegrityOpts,
+    inner: Inner
+}
+
+impl<R: Read> Read for SSRIStream<R> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.inner.read(buf).and_then(|size| {
+            self.builder.input(&buf[0..size]);
+            Ok(size)
+        })
+    }
+}
+
+impl<W: Write> Write for SSRIStream<W> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.inner.write(buf).and_then(|size| {
+            self.builder.input(&buf[0..size]);
+            Ok(size)
+        })
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl<R> SSRIStream<R> {
+    fn into_inner(self) -> (ssri::Integrity, R) {
+        (self.builder.result(), self.inner)
+    }
+}
+
+/// Take a stream of file-like entries and write them as a "packfile" for
+/// fast random access to members in the future.
+pub async fn write_entries<P, I, D>(cache: P, mut entry_stream: D) -> Result<Integrity>
+where
+    P: AsRef<Path>,
+    I: FileLike,
+    D: futures::stream::Stream<Item = I> + std::marker::Unpin + Send + Sync + 'static {
+    let mut entry_hash = std::collections::HashMap::new();
+    let mut dest = SSRIStream {
+        inner: std::io::Cursor::new(Vec::with_capacity(1024 * 1024 * 10)),
+        builder: ssri::IntegrityOpts::new().algorithm(ssri::Algorithm::Sha256)
+    };
+    let mut pb = cache.as_ref().to_owned();
+
+    async_std::task::spawn(async move {
+        let mut offsets = std::collections::BTreeMap::new();
```

Comment:
This is a map from "integrity bytes" to "offset in the packfile". We use a BTreeMap here so that the keys are automatically lexically sorted for us on insertion.